components/BottomNavigation/src/MDCBottomNavigationBarController.m

```diff
@@ -148,6 +160,47 @@ - (BOOL)bottomNavigationBar:(MDCBottomNavigationBar *)bottomNavigationBar
   return YES;
 }
 
+#pragma mark - Key Value Observation Methods
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSKeyValueChangeKey, id> *)change
+                       context:(void *)context {
+  if (context != kObservationContext) {
+    [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
+    return;
+  }
+
+  id newValue = [change objectForKey:NSKeyValueChangeNewKey];
+  if (object == self.navigationBar &&
+      [keyPath isEqualToString:NSStringFromSelector(@selector(items))] &&
+      [newValue isKindOfClass:[NSArray class]]) {
+    [self didUpdateNavigationBarItemsWithNewValue:(NSArray *)newValue];
+  }
+}
+
+- (void)didUpdateNavigationBarItemsWithNewValue:(NSArray *)items {
+  // Verify tab bar items correspond with the view controllers tab bar items.
+  BOOL isItemViewControllerCountEqual = items.count == self.viewControllers.count;
+  NSAssert(isItemViewControllerCountEqual, [self navigationBarItemsChangedExceptionDescription]);
+  if (!isItemViewControllerCountEqual) {
+    NSLog(@"%@", [self navigationBarItemsChangedExceptionDescription]);
+    return;
+  }
+
+  // Verify each new and the view controller's tab bar items are equal.
+  for (NSUInteger i = 0; i < self.viewControllers.count; i++) {
+    UITabBarItem *viewControllerTabBarItem = [self.viewControllers objectAtIndex:i].tabBarItem;
+    UITabBarItem *newTabBarItem = [items objectAtIndex:i];
+    BOOL tabBarItemsEqual = [viewControllerTabBarItem isEqual:newTabBarItem];
+    NSAssert(tabBarItemsEqual, [self navigationBarItemsChangedExceptionDescription]);
```

Comment:
We spoke offline - let's change the `navigationBarItemsChangedExceptionDescription` to return an NSException and have it raised here. This will match UITabBarController's behavior.